src/core/connection_pool.c

```diff
@@ -0,0 +1,350 @@
+            sizeof(HQUIC) * NumberOfConnections);
+        goto Error;
+    }
+
+    //
+    // Start testing ports and creating connections.
+    //
+
+    const uint32_t ConnectionsPerProc = (NumberOfConnections / RssProcessorCount) + 1;
+
+    for (uint32_t i = 0; i < NumberOfConnections; i++) {
+        uint32_t RssProcIndex;
+        do {
+            QuicAddrIncrementPort(&LocalAddress);
+            uint32_t RssHash = 0, Offset;
+            //
+            // Calculate the Toeplitz Hash as if receiving packets from the
+            // ResolvedRemoteAddress to find the RSS processor.
+            //
+            CxPlatToeplitzHashComputeRss(
+                &ToeplitzHash,
+                &ResolvedRemoteAddress,
+                &LocalAddress,
+                &RssHash,
+                &Offset);
+            uint32_t Mask = (RssConfig->RssIndirectionTableLength / sizeof(CXPLAT_PROCESSOR_INFO)) - 1;
+            RssProcIndex = RssProcessorCount;
+            CXPLAT_PROCESSOR_INFO* RssProc = &RssConfig->RssIndirectionTable[RssHash & Mask];
+            for (uint32_t j = 0; j < RssProcessorCount; j++) {
+                #pragma prefast(suppress:6385, "SAL doesn't understand this is safe");
+                if (RssProcessors[j].Group == RssProc->Group && RssProcessors[j].Index == RssProc->Index) {
+                    RssProcIndex = j;
+                    break;
+                }
+            }
+            CXPLAT_DBG_ASSERT(RssProcIndex < RssProcessorCount);
+            if (ConnectionCounts[RssProcIndex] >= ConnectionsPerProc) {
+                //
+                // This processor already has enough connections on it, so try another port number.
+                //
+                continue;
+            }
+
+            CXPLAT_SOCKET* TestSocket = NULL;
+            CXPLAT_UDP_CONFIG TestConfig = {
+                .LocalAddress = &LocalAddress,
+                .RemoteAddress = &ResolvedRemoteAddress
+            };
+            Status = CxPlatSocketCreateUdp(MsQuicLib.Datapath, &TestConfig, &TestSocket);
+            if (QUIC_FAILED(Status)) {
+                //
+                // This port is already in use. Gotta try another port number.
+                //
+                continue;
+            }
+
+            //
+            // Making it this far means we can create the connection!
+            //
+            CxPlatSocketDelete(TestSocket);
+            break;
+
+        } while (TRUE);
+
+        Status = MsQuicConnectionOpen(Registration, Handler, Context, &Connections[i]);
+        if (QUIC_FAILED(Status)) {
+            QuicTraceLogError(
+                ConnPoolOpenConnection,
+                "[conp] Failed to open connection[%u], 0x%x",
+                i,
+                Status);
+            goto Error;
+        }
+        CreatedConnections++;
+
+        Status = MsQuicSetParam(Connections[i], QUIC_PARAM_CONN_REMOTE_ADDRESS, sizeof(ResolvedRemoteAddress), &ResolvedRemoteAddress);
```

Comment:
Use `QuicConnSetParam` instead. Also, you can debug assert the result, because it can only fail if you mess up the code here (i.e. invalid parameters). Same for local address.

Reply:
I don't want to only assert in case there's a change in behavior in the future. I don't want users of the API to be doing the wrong thing.
But I will add an assert that this shouldn't fail in our test runs so we can catch behavior changes.